package.json

```diff
@@ -3,6 +3,21 @@
   "title": "jQuery",
   "description": "JavaScript library for DOM operations",
   "version": "4.0.0-pre",
+  "type": "commonjs",
+  "exports": {
+    ".": {
+      "node": "./dist/jquery.js",
```

Comment:
The `"node"` field should only be used when an export is _only_ usable by Node.js, for example if it contains compiled native extensions. In general we encourage packages to define only `require` and `default` whenever possible, to maximize compatibility.

My previous comment https://github.com/jquery/jquery/pull/5122#discussion_r976844183 is still valid; see https://github.com/GeoffreyBooth/js-mjs-mime-type-test. Even in 2023 it’s unsafe to assume that web servers and CDNs will serve `.mjs` files correctly. Your `default` export should be a `.js` file.

Since the most important files, under `src`, are written in ESM with `.js` extensions, I would treat `.js` as meaning “ESM” throughout the codebase for consistency. Therefore the root `package.json` would get `"type": "module"` and all the CommonJS files scattered around would get `.cjs` extensions. Then this block becomes:

```js
  "exports": {
    ".": {
       "require": "./dist/jquery.cjs",
       "default": "./dist/jquery.js"
```

And likewise for `"./slim"` and so on.

Reply:
@GeoffreyBooth

Responding to each aspect:

## The `node` field in `exports`

If I only define `require` and `default` I will hit dual package hazard. Per [Node.js packages docs](https://nodejs.org/api/packages.html#dual-package-hazard), to avoid it one can isolate the state in a CommonJS module and require it in an ESM. This is Node-specific (not suitable for pure ESM envs) so I’d have to add the `node`.

This is what I actually initially aimed at, but then I realized that in our case the ESM module would just have a default export matching `module.exports` from the CommonJS version so the ESM wrapper would be useless, Node is properly serving our CommonJS module to ESM consumers fine by default.

The `node` field is Node-specific in our case, in that it makes Node always serve the CommonJS version to avoid dual package hazard.

## `.mjs`

As for `.mjs`, why is serving the CommonJS version with the `.cjs` extension less risky?

Also, right now we produce the same files for browser & Node. The CommonJS version is also the default script version for the browser, it having a `.cjs` extension would be very unintuitive. On the other hand, both ESM & CommnJS files having the same `.js` extension won’t work for Node… It seems that avoiding `.mjs` would require keeping the CommonJS file duplicated, with both the `.js` and `.cjs` extension. 😕

Reply:
Or maybe the solution to avoid `.mjs` and `.cjs` is to keep them in separate `dist` & `dist-esm` directories and add proper `package.json` with the `type` fields?

Reply:
The dual package hazard is only an issue if a package contains internal state. Does `jQuery` contain any state? Do people register plugins on the `jQuery` global perhaps, and therefore would be surprised if they existed on `jQuery` in one module environment but not the other?

Assuming that that _is_ an issue, then you need to choose between these options:
- Ship the CommonJS version to all Node users, regardless of whether the Node application is written in CommonJS or ESM. This might lead to bundlers including _all_ of jQuery rather than tree-splitting the ESM version, and it might annoy the developer who’s writing ESM to see CommonJS in the stack trace. Neither of these are showstoppers, but they’re worth considering.
- Ship only the ESM version, period, joining the list of packages that have dropped CommonJS support in their latest major versions. Obviously this would greatly annoy any Node developers writing CommonJS applications.
- Ship both versions where Node ESM users get the ESM version and Node CommonJS users get the CommonJS version. You would probably have to add a warning in the docs telling people not to add custom properties to the `jQuery` object or other exports, that they expect to see across module systems.

If the hazard is a real concern, then sure, maybe the first option is the least bad one.

> As for `.mjs`, why is serving the CommonJS version with the `.cjs` extension less risky?

Because web servers and CDNs aren’t serving the CommonJS version; that version is only for Node or bundlers, who would be loading it from files (and which all have support for the `.cjs` extension; _all_ Node versions treat unknown extensions as CommonJS, so it works back to the beginning). Whereas the version for use in browsers needs to be served with the correct MIME type, and per https://github.com/GeoffreyBooth/js-mjs-mime-type-test even in 2023 many big open source webservers like Nginx and Apache don’t serve `.mjs` with the correct MIME type by default. Or put another way, there’s no benefit to `.mjs` but there’s a big downside to this potential incompatibility.

> right now we produce the same files for browser & Node

Aren’t you producing both a browser ESM version and a browser AMD version? So it’s the browser AMD version that is identical to the Node CommonJS version?

Yes, you could just output it twice, like `jquery.amd.js` and `jquery.cjs`, and the contents of the two versions happen to be identical. That wouldn’t be terrible. It probably would be preferable for them to _not_ be identical, as the `.cjs` version doesn’t need some of the AMD stuff like checking if `require` is defined; it can just assume that it’s in a CommonJS environment.

You could also do something like put this combined AMD/CommonJS file in a subfolder like `dist/browser-amd-and-node-commonjs/jquery.js` and add `dist/browser-amd-and-node-commonjs/package.json` with `"type": "commonjs"`. Then it’s a little more explicit that this file is being reused. Personally I would just do the first option, outputting it twice, so that you have the opportunity to optimize each one a bit rather than having checks that are only needed since you’re reusing the same file in two very different environments.

Reply:
@GeoffreyBooth 
> The dual package hazard is only an issue if a package contains internal state. Does `jQuery` contain any state?

Yes. jQuery plugins defining methods on jQuery prototype are one thing (this is a common pattern that we cannot really change at this point) but there's also the `.data()` method that acts as a data storage per DOM node; each jQuery instance has a separate storage and it definitely can cause issues if you suddenly have two jQuery instances on the page where you expect one, not to mention JS size concerns.

> Assuming that that _is_ an issue, then you need to choose between these options:
> 
> * Ship the CommonJS version to all Node users, regardless of whether the Node application is written in CommonJS or ESM. This might lead to bundlers including _all_ of jQuery rather than tree-splitting the ESM version, and it might annoy the developer who’s writing ESM to see CommonJS in the stack trace. Neither of these are showstoppers, but they’re worth considering.

The ESM version in the PR is almost identical to the CommonJS one, differing mostly in that the former assigns `jQuery` to `module.exports` and the latter does `export default jQuery`.

There's no tree-shaking involved. First of all, we can't mark jQuery as side-effects free via the `package.json` `"sideEffects": false` flag; secondly, different jQuery modules still attach themselves to jQuery or its prototype. This is a fundamental property of jQuery that we cannot change without introducing enormous breaking changes that are impractical at this stage of the project.

For people wanting to only include a part of jQuery, we have two solutions:
* one independent of `exports` or even npm - compile a custom version of jQuery that excludes some modules; we document that in README
* one npm-based - to import modules from `jquery/src/**` directly

Both are independent from the ESM versions of `jquery` or `jquery/slim`.

As for bundlers, I thought they don't apply the `node` condition - but even if they were, I guess that'd be fine as long as they'd do it consistently and as long as it doesn't require any special config to understand CommonJS. I think this is true for Webpack but I'll need to verify other bundlers.

> * Ship only the ESM version, period, joining the list of packages that have dropped CommonJS support in their latest major versions. Obviously this would greatly annoy any Node developers writing CommonJS applications.

We still need a script version that sets the `jQuery` & `$` globals for the browser - and right now that same file is also used to export as CommonJS. So this would at least not reduce the number of files we need to generate in the build & publish.

I think CommonJS won't go away any time soon so I'd be afraid for jQuery to make such a huge break for CommonJS users, especially for a library that is not bleeding-edge.

> * Ship both versions where Node ESM users get the ESM version and Node CommonJS users get the CommonJS version. You would probably have to add a warning in the docs telling people not to add custom properties to the `jQuery` object or other exports, that they expect to see across module systems.

This wouldn't fly for the reasons stated above.

> > As for `.mjs`, why is serving the CommonJS version with the `.cjs` extension less risky?
> 
> Because web servers and CDNs aren’t serving the CommonJS version; that version is only for Node or bundlers, who would be loading it from files (and which all have support for the `.cjs` extension; _all_ Node versions treat unknown extensions as CommonJS, so it works back to the beginning).

Not in our case; we use the same file as the default `jquery.js` script for browser consumption.

> Whereas the version for use in browsers needs to be served with the correct MIME type, and per [GeoffreyBooth/js-mjs-mime-type-test](https://github.com/GeoffreyBooth/js-mjs-mime-type-test) even in 2023 many big open source webservers like Nginx and Apache don’t serve `.mjs` with the correct MIME type by default. Or put another way, there’s no benefit to `.mjs` but there’s a big downside to this potential incompatibility.

Makes sense, that definitely means we cannot serve `.mjs` to browsers.

> > right now we produce the same files for browser & Node
> 
> Aren’t you producing both a browser ESM version and a browser AMD version? So it’s the browser AMD version that is identical to the Node CommonJS version?

Registering as an AMD module is currently done in all files, conditional on the `define` check. This is a small code snippet so I didn't want to generate yet another file just for that. It may be a good idea to skip that code in the ESM version, though.

> Yes, you could just output it twice, like `jquery.amd.js` and `jquery.cjs`, and the contents of the two versions happen to be identical. That wouldn’t be terrible. It probably would be preferable for them to _not_ be identical, as the `.cjs` version doesn’t need some of the AMD stuff like checking if `require` is defined; it can just assume that it’s in a CommonJS environment.

Since we reuse the CommonJS version as the jQuery script one, I think I'm OK with continuing to reuse that for AMD registration as well.

> You could also do something like put this combined AMD/CommonJS file in a subfolder like `dist/browser-amd-and-node-commonjs/jquery.js` and add `dist/browser-amd-and-node-commonjs/package.json` with `"type": "commonjs"`. Then it’s a little more explicit that this file is being reused. Personally I would just do the first option, outputting it twice, so that you have the opportunity to optimize each one a bit rather than having checks that are only needed since you’re reusing the same file in two very different environments.

jQuery gets published to npm, bower, to three different CDNs (using different naming schemes) & used in so many ways that I'd rather not mutliply the number of built files if there's no big enough gain in that. In this PR, we're already generating 4 versions: ESM/script, full/slim build and each combination of them.

I like the option with separate folders, I'll try that.

Reply:
Just to help me keep things straight, can you list the various output files and the intended consumers of each? (Leaving aside `slim` since I assume that’ll just be a copy of whatever pattern you use for the main version.)

Reply:
Disregarding slim (you're right, it should mostly duplicate what the regular version does), we're generating the following files right now in the PR:
* `jquery.js`: meant to be consumed by:
    * browsers via `<script src="jquery.js">`, sets the `jQuery` & `$` properties on `window`
    * browsers via AMD - still sets the `jQuery` & `$` properties on `window` as well
    * in a CommonJS environment: if a global `window` is present at the moment of code evaluation, it exports (via `module.exports = jQuery`) a jQuery instance but does not set any globals. If `window` is not present when `require`'d, it instead exports a factory function accepting a `window` shim (e.g. generated with JSDOM) and returning a jQuery instance, setting `jQuery` & `$` properties on that `window` instance.
* `jquery.min.js`: a minified version of `jquery.js`; no other changes
* `jquery.min.map`: a source map file for `jquery.min.js`
* `jquery.mjs` (I know I'll need to change the extension): meant to be consumed by:
    * browsers via `<script type="module">`; does not set any globals
    * in a non-browser ESM environment: if a global `window` is present at the moment of code evaluation, it exports (via `export default`) a jQuery instance but does not set any globals. If `window` is not present when `require`'d, it instead exports a factory function accepting a `window` shim (e.g. generated with JSDOM) and returning a jQuery instance, setting `jQuery` & `$` properties on that `window` instance.

I'm considering removing this weird logic of when to set the `jQuery` & `$` properties on `window` in CommonJS/ESM. I'm also thinking about extracting the factory export mode to separate entry points as it's pretty confusing, although that'd duplicate the number of entry points (we'd need something similar for slim as well).

Looking at [`src/wrapper.js`](https://github.com/mgol/jquery/blob/b538327b89a458773bf51c5fd9e484ed22263680/src/wrapper.js) and [`src/wrapper-esm.js`](https://github.com/mgol/jquery/blob/b538327b89a458773bf51c5fd9e484ed22263680/src/wrapper-esm.js) may help with understanding the logic. `src/wrapper.js` is used to generate `jquery.js`, `src/wrapper-esm.js` - to generate `jquery.mjs`.

As for `jquery.min.js`: I haven't included it in `exports`, expecting most usage to come through the CNDs, loaded directly into a script tag. Are there any use cases I may be missing here?

Reply:
> As for `jquery.min.js`: I haven’t included it in `exports`, expecting most usage to come through the CNDs, loaded directly into a script tag. Are there any use cases I may be missing here?

The testing use case, using something like JSDom, should presumably be using the minified jQuery as it’s more realistic to what would be run in a browser. Ditto for server-side rendering. It shouldn’t make a difference, but purists might care.

I wouldn’t expect the ESM version to be setting globals. If you haven’t shipped that yet, perhaps it can just provide a `jQuery` named export and nothing more, and if the user wants it set on `globalThis` then the user can do so?

You should also arguably be shipping a minified browser ESM version, not just a minified browser script version. Users referencing it via `script type="module"` or a direct `import` statement would presumably want the minified version.

I think you could distinguish all these versions either via filename part segments or via folders. So something like (excluding the `.map` files):

- `jquery.amd.js`: for browsers via regular `script` or via AMD
- `jquery.amd.min.js`: for browsers via regular `script` or via AMD, minified
- `jquery.cjs`: for Node CommonJS
- `jquery.esm.js`: for browsers via `script type="module"` or `import`, or Node ESM
- `jquery.esm.min.js`: for browsers via `script type="module"` or `import`, minified, or Node ESM

or with subfolders:

- `script/jquery.js`: for browsers via regular `script` or via AMD, or for Node CommonJS
- `script/jquery.min.js`: for browsers via regular `script` or via AMD, or for Node CommonJS
- `script/package.json`: file with `{ "type": "commonjs" }` to start a new scope since the root would be `"type": "module"`
- `module/jquery.js`: for browsers via `script type=”module”` or `import`, or Node ESM
- `module/jquery.min.js`: for browsers via `script type=”module”` or `import`, or Node ESM, minified

Reply:
I haven’t added minifying for ESM builds because we use UglifyJS which is ES5-only. Switching to Terser increases the final size a lot so we can’t do it for the script build unless Terser improves significantly. I was thinking about tackling that in a future minor update as that shouldn’t require breaking changes.

Regardless of the above, if we were to expose at least the current minified files via `exports`, what would be the best way to do it?

Reply:
> I haven’t added minifying for ESM builds because we use UglifyJS which is ES5-only.

You could also use `esbuild` to minify.

> Regardless of the above, if we were to expose at least the current minified files via `exports`, what would be the best way to do it?

Well I think first you need to decide which of my two suggested `dist` file/folder structures you prefer. Then we can come up with an `"exports"` definition for it.

Reply:
### File/directory structure

I think I prefer the second option, with subfolders. We will need to send these files to CDNs as well so the more uniformity we have, the better.

The only change I'd make is to also include the info about the file type in the name so that for CDN purposes we can flatten the directory structure without changing the file names, and so that it's easier to associate which file is which. So, modifying your proposal:
- `script/jquery.js`: for browsers via regular `script` or via AMD, or for Node CommonJS
- `script/jquery.min.js`: for browsers via regular `script` or via AMD, or for Node CommonJS
- `script/package.json`: file with `{ "type": "commonjs" }` to start a new scope since the root would be `"type": "module"`
- `module/jquery.module.js`: for browsers via `script type="module"` or `import`, or Node ESM
- `module/jquery.module.min.js`: for browsers via `script type="module"` or `import`, or Node ESM, minified

(optionally, we may decide to rename `jquery.module.js` to something shorter like `jquery.esm.js` but it won't change the gist of the proposal)

### Minifiers

I did a local comparison of our current UglifyJS setup with Terser & esbuild. The commands used:
```
npx uglify-js@3.4.7 dist/jquery.js --output dist/jquery.uglify.min.js --compress hoist_funs=false,loops=false --beautify beautify=false,ascii_only=true,preamble='"/*! jQuery v4.0.0-pre | (c) OpenJS Foundation and other contributors | jquery.org/license */"' --mangle && echo >>dist/jquery.uglify.min.js

npx terser@5.17.6 dist/jquery.js --output dist/jquery.terser.min.js --compress hoist_funs=false,loops=false --format beautify=false,ascii_only=true,ecma=2015,comments=false,preamble='"/*! jQuery v4.0.0-pre | (c) OpenJS Foundation and other contributors | jquery.org/license */"' --mangle && echo >>dist/jquery.terser.min.js

cat dist/jquery.js | npx esbuild@0.17.19 --minify >dist/jquery.esbuild.min.js
```

esbuild doesn't have the preamble option so I also manually replaced it with the shorter version meant for the minified files. Results were as follows (sizes in bytes):

| Minifier | min size | min+gzip size | min delta | min+gzip delta |
|----------|----------|---------------|-----------|----------------|
| UglifyJS |   78,926 |        27,537 |        +0 |             +0 |
| Terser   |   79,251 |        27,610 |      +325 |            +73 |
| esbuild  |   79,556 |        28,464 |      +630 |           +927 |

esbuild seems particularly not optimized for gzip so we'd only really consider Terser. +73 gzipped may be acceptable. Otherwise, I could consider minifying the whole file without the export pragma and then attaching it but with +73 bytes I'm tempted to just bite the bullet & migrate. I think the difference was larger in the past.

Reply:
`esbuild` has a `--banner` option which is their version of `--preamble`: https://esbuild.github.io/api/#banner. But yeah if Terser can generate an ESM minified file that’s smaller, sure, go with that.

Webpack has some good examples of various `"exports"` configurations. This one looks like your case: https://webpack.js.org/guides/package-exports/#providing-commonjs-and-esm-version-stateful

So using your proposed structure and based on https://webpack.js.org/guides/package-exports/#example-1, maybe something like this (just for the main version for now, and slim will be a copy):

```json
{
  "type": "module",
  "exports": {
    "node": {
      "development": {
        "module": "./dist/module/jquery.module.js",
        "import": "./dist/module-wrapper.js",
        "require": "./dist/script/jquery.js"
      },
      "production": {
        "module": "./dist/module/jquery.module.min.js",
        "import": "./dist/module-wrapper.min.js",
        "require": "./dist/script/jquery.min.js"
      },
      "default": "./dist/commonjs-wrapper.cjs"
    },
    "development": "./dist/module/jquery.module.js",
    "production": "./dist/module/jquery.module.min.js",
    "script": "./dist/script/jquery.min.js",
    "default": "./dist/module/jquery.module.min.js"
  }
}
```

With wrapper files like this:

```js
// dist/module-wrapper.js
export * from "./script/jquery.js";

// dist/module-wrapper.min.js
export * from "./script/jquery.min.js";

// dist/commonjs-wrapper.cjs
module.exports = require("./script/jquery.js");
```

I put these one level up because they’re just wrapper files, and also only intended for Node; but you can move them wherever they make the most sense to you (and you can move `dist/commonjs-wrapper.cjs` into `script` if you want to avoid the `.cjs` extension, but I think it’s kind of appropriate here because this file really is _only_ for Node CommonJS).

These wrappers all assume you only want to provide a default export. You could also provide both a default and named exports. For example, for an ESM wrapper:

```js
import package from "./script/jquery.js";
export default package;
export const { jQuery, $ } = package;
```

Or from the CommonJS wrapper:

```js
const package = require("./script/jquery.js");
module.exports = package;
module.exports.jQuery = package.jQuery;
module.exports.$ = package.$;
```

The CommonJS wrapper has an explicit `module.exports.jQuery` so that it triggers Node’s CommonJS named exports detection, so that Node ESM users can write `import { jQuery } from "jquery"` and it will work, even though the export is a CommonJS file.

It’s uncommon to use minified files with Node so I left the unminified version as the default for Node users; but the minified one for others, so that bundlers pick the minified build by default.

Reply:
That’s what I don’t get: what’s the point of the Node module wrapper assuming I only want a default export? The wrapper depends on CommonJS `module.exports` being turned into a default export in an ESM consumer anyway, why not use it directly & reduce the complexity?

Reply:
> That’s what I don’t get: what’s the point of the Node module wrapper assuming I only want a default export?

Many people (including me) frown on default exports. They’re explicitly discouraged in https://webpack.js.org/guides/package-exports/#guidelines and there are linting rules to enforce not using them. You can’t drop the default export at this point because of backward compatibility, but I would at least ship named exports in addition to it.

Reply:
> wouldn’t that mean we have to attach them to the jQuery object, making us have `jQuery.jQuery === jQuery` & `jQuery.$ === jQuery`?

So I did some testing. All of these assertions return true.

```shell
mkdir test && cd test
npm install jquery
touch wrapper.cjs consumer.cjs wrapper.mjs consumer.mjs
```

```js
// wrapper.cjs
const pkg = require("jquery");
module.exports = pkg;
module.exports.jQuery = pkg;
module.exports.$ = pkg;

// consumer.cjs
const wrapped = require("jquery");
const wrapper = require("./wrapper.cjs");
console.assert(wrapper === wrapped, 1);
console.assert(wrapper.jQuery === wrapped, 2);
console.assert(wrapper.jQuery.jQuery === wrapped, 3);

// wrapper.mjs
import pkg from "jquery";
export default pkg;
export const jQuery = pkg;
export const $ = pkg;

// consumer.mjs
import wrapped from "jquery";
import wrapper, { jQuery } from "./wrapper.mjs";
console.assert(wrapper === wrapped, 1);
console.assert(jQuery === wrapped, 2);
console.assert(jQuery.jQuery === undefined, 3);
```

So you’re right _for the CommonJS wrapper,_ because that’s how CommonJS works, but not for the ESM wrapper. The `module.exports.jQuery = pkg;` line not only creates the `jQuery` module export but also adds a `jQuery` property to the `jQuery` object, because that’s what `module.exports` has already been assigned to be. I guess the question is whether this is a problem; they’re all references to the same object, so we don’t have a hazard.

Looking again at the Webpack example, their CommonJS example was only to serve production versus development versions based on `process.env.NODE_ENV` as a fallback for when the user isn’t using `--conditions`. I don’t think we need to do that; we can just serve the unminified version as our default for Node, and the minified one can be enabled only via `--conditions=production`. Therefore we wouldn’t need a CommonJS wrapper at all:

```json
{
  "type": "module",
  "exports": {
    "node": {
      "production": {
        "module": "./dist/module/jquery.module.min.js",
        "import": "./dist/module-wrapper.min.js",
        "require": "./dist/script/jquery.min.js"
      },
      "development": {
        "module": "./dist/module/jquery.module.js",
        "import": "./dist/module-wrapper.js",
        "require": "./dist/script/jquery.js"
      },
      "module": "./dist/module/jquery.module.js",
      "import": "./dist/module-wrapper.js",
      "require": "./dist/script/jquery.js"
    },
    "production": "./dist/module/jquery.module.min.js",
    "development": "./dist/module/jquery.module.js",
    "script": "./dist/script/jquery.min.js",
    "default": "./dist/module/jquery.module.min.js"
  }
}
```

```js
// module-wrapper.js (and module-wrapper.min.js is identical other than importing jquery.min.js)
import pkg from "./script/jquery.js";
export default pkg;
export const { jQuery, $ } = pkg;
```

With this approach, we can get both the named exports for ESM consumers without adding new properties on the `jQuery` object.

Reply:
Do you think it's acceptable for the script a.k.a. CommonJS version to just export jQuery via `module.exports = jQuery` and only expose named `jQuery` & `$` exports in the ESM version?

I had the impression the recommended behavior is to have both expose the same tokens but maybe I'm wrong.

A few other remarks:
* I think I'd prefer to have two top-level directories: `dist` & `dist-module` so that the script version remains available at `dist/jquery.js` & `dist/jquery.min.js` - there is a lot of code out there that loads those files directly from the jQuery npm folder, bypassing the `require` & `import` interface
* Do we need the `node.development` section when its contents are already avaiable directly at the `node` level?
* There are `production` & `development` conditions at the top level, pointing to the module version - but the `script` version only has the minified file. What's the reason for this inconsistency?

Reply:
> Do you think it’s acceptable for the script a.k.a. CommonJS version to just export jQuery via `module.exports = jQuery` and only expose named `jQuery` & `$` exports in the ESM version?

The ESM wrapper wouldn’t _only_ expose the named exports. The default _and_ named exports should work:

```js
// Both of these should work:
import jQuery from "jquery";
import { jQuery, $ } from "jquery";
```

> * Do we need the `node.development` section when its contents are already avaiable directly at the `node` level?

Probably not, but it feels weird not to list it? Up to you.

> * There are `production` & `development` conditions at the top level, pointing to the module version - but the `script` version only has the minified file. What’s the reason for this inconsistency?

`"script"` is something I saw in the Webpack docs; it’s a condition only for bundlers. Node ignores it. It’s for bundlers generating output to go into an HTML `script` tag (and not a `script type="module"` tag). I assume that most of the time if we’re putting jQuery into a `script` tag, it should be the minified version.

Reply:
> The ESM wrapper wouldn’t only expose the named exports. The default and named exports should work:

Yes, I know. My question was whether it's acceptable for the CommonJS version to only have the "default export" in the form of `module.exports = jQuery` and the ESM version having both the default & named ones. So far, I had an impression libraries try to maintain a compatible interface between those two versions.

Reply:
> Yes, I know. My question was whether it’s acceptable for the CommonJS version to only have the “default export” in the form of `module.exports = jQuery` and the ESM version having both the default & named ones. So far, I had an impression libraries try to maintain a compatible interface between those two versions.

This is a limitation of CommonJS because you’re using default exports. You have to choose between supplying only the default export _or_ attaching extra properties like `jQuery.jQuery`. I maintain CoffeeScript and I chose the latter:

https://github.com/jashkenas/coffeescript/blob/8984dd3eaf8c9c6c18c3c3400134bd32e01d4864/src/index.coffee#L140-L160

No one’s ever complained to me that when they did `const CoffeeScript = require('coffeescript')` that there was an extra `CoffeeScript.CoffeeScript` property that they could access (or that it was missing for ESM consumers). It’s a circular reference, so you don’t need to worry about it having different state than the root object.

There are two versions of “keeping the API the same.” There’s the shape of what gets exported, and the signatures of how to import things. The `jQuery.jQuery` version allows for the following:

```js
import { jQuery } from 'jquery';
// in a separate CommonJS file:
const { jQuery } = require('jquery');
```

Whereas if CommonJS only supplies the default export, the second line could only be written as `const jQuery = require('jquery');`. Conversely if you supply only the default export to _both_ consumers, ESM consumers would be forced to write `import jQuery from 'jquery';` which goes against best practices of avoiding default exports. Most of your consumers are likely to be writing ESM, so I would bias things toward them.

> currently, in your proposal `node` entries have the `module`, `import` & `require` conditions. Does it matter that `default` is not specified or is it better to replace the last entry on the list with `default`?

The `module` condition is only for bundlers building for Node; Node itself ignores it. That just leaves `import` and `require`, and we know that Node will choose one or the other, so there’s not really any reason to add anything else. You can add `default` for completeness but it’ll never get used by either Node or bundlers.

Reply:
@GeoffreyBooth I wonder - do we need to provide a default export at all? The CommonJS version would have one but if we keep different exported tokens between CommonJS & ESM versions anyway, can we just go all the way and only export named `$` & `jQuery`?

Reply:
> do we need to provide a default export at all?

For backward compatibility, I would assume you need to provide it; aren't people writing `import jQuery from 'jquery'` today? And you're forced to support that if you support Node via the CommonJS wrapper approach; so then _not_ allowing that syntax in pure ESM environments would seem likely to cause confusion.

Reply:
Ah, right, I forgot about a lot of the code relying on bundlers that would auto-switch to the ESM version & break.

Reply:
@GeoffreyBooth I'm working on updates right now. I'm missing why we need the `production` section for the `node` condition. I understand why it's needed outside of Node.js and so that we need the `production`, `development` & `module` fields there but why in Node?